src/flow/src/transform/plan.rs

```diff
@@ -302,9 +203,110 @@ impl TypedPlan {
+        .clone()
+        .into_iter()
+        .map(|i| {
+            if i < key_val_plan.key_plan.input_arity {
+                ScalarExpr::Column(i)
+            } else {
+                key_val_plan.key_plan.expressions[i - key_val_plan.key_plan.input_arity].clone()
+            }
+        })
+        .collect_vec();
+    let mut shift_offset = 0;
+    let mut shuffle: BTreeMap<usize, usize> = BTreeMap::new();
+    let special_keys = key_exprs
+        .clone()
+        .into_iter()
+        .enumerate()
+        .filter(|(idx, p)| {
+            shuffle.insert(*idx, *idx + shift_offset);
+            if matches!(
+                p,
+                ScalarExpr::CallUnary {
+                    func: UnaryFunc::TumbleWindowFloor { .. },
+                    ..
+                } | ScalarExpr::CallUnary {
+                    func: UnaryFunc::TumbleWindowCeiling { .. },
+                    ..
+                }
+            ) {
+                if matches!(
+                    p,
+                    ScalarExpr::CallUnary {
+                        func: UnaryFunc::TumbleWindowFloor { .. },
+                        ..
+                    }
+                ) {
+                    shift_offset += 1;
+                }
+                true
+            } else {
+                false
+            }
+        })
+        .collect_vec();
+    let spec_key_arity = special_keys.len();
+    if spec_key_arity == 0 {
+        return Ok(());
+    }
+
+    // shuffle proj_exprs
+    // because substrait use offset while assume `tumble` only return one value
+    for proj_expr in proj_exprs.iter_mut() {
+        proj_expr.expr.permute_map(&shuffle)?;
+    } // add key to the end
+    for (key_idx, _key_expr) in special_keys {
+        // here we assume the output type of reduce operator(`reduce_output_type`) is just first keys columns, then append value columns
+        // so we can use `key_idx` to index `reduce_output_type` and get the keys we need to append to `proj_exprs`
+        proj_exprs.push(
+            ScalarExpr::Column(key_idx)
+                .with_type(reduce_output_type.typ().column_types[key_idx].clone()),
+        );
+    }
+
+    // check if normal expr in group exprs are all in proj_exprs
+    let all_cols_ref_in_proj: BTreeSet<usize> = proj_exprs
+        .iter()
+        .filter_map(|e| {
+            if let ScalarExpr::Column(i) = &e.expr {
+                Some(*i)
+            } else {
+                None
+            }
+        })
+        .collect();
+    for (key_idx, key_expr) in key_exprs.iter().enumerate() {
+        if let ScalarExpr::Column(_) = key_expr {
+            if !all_cols_ref_in_proj.contains(&key_idx) {
+                return InvalidQuerySnafu{
+                    reason: format!("Expect normal column in group by also appear in projection, but column {}(name is {}) is missing", key_idx, reduce_output_type.get_name(key_idx).clone().map(|s|format!("'{}'",s)).unwrap_or("unknown".to_string()))
```

Comment:
sugg: write this into two statements. `format!` macro cannot be formatted properly